## Promise API
There are 6 static methods in the Promise class.

``Promise.all ``takes an iterable (usually, an array of promises) and returns a new promise.
The new promise resolves when all listed promises are resolved, and the array of their results becomes its result.
The syntax is:
``
let promise = Promise.all(iterable);
``
Goal of this is that we want many promises to execute in parallel and wait until all of them are ready.
``Promise.all`` takes an iterable (usually, an array of promises) and returns a new promise.
For instance, the ``Promise.all`` below settles after 3 seconds, and then its result is an`` array [1, 2, 3]``:
```js
Promise.all([
  new Promise(resolve => setTimeout(() => resolve(1), 3000)), // 1
  new Promise(resolve => setTimeout(() => resolve(2), 2000)), // 2
  new Promise(resolve => setTimeout(() => resolve(3), 1000))  // 3
]).then(alert); // 1,2,3 when promises are ready: each promise contributes an array member
```

A common trick is to map an array of job data into an array of promises, and then wrap that into ``Promise.all``.
```js
let urls = [
  'https://api.github.com/users/iliakan',
  'https://api.github.com/users/remy',
  'https://api.github.com/users/jeresig'
];

// map every url to the promise of the fetch
let requests = urls.map(url => fetch(url));

// Promise.all waits until all jobs are resolved
Promise.all(requests)
  .then(responses => responses.forEach(
    response => alert(`${response.url}: ${response.status}`)
  ));
```
**In case of an error, other promises are ignored**

- we are able to pass ready values to Promise.all where convenient.
```js
Promise.all([
  new Promise((resolve, reject) => {
    setTimeout(() => resolve(1), 1000)
  }),
  2,
  3
]).then(alert); // 1, 2, 3
```

### Promise.allSettled
``Promise.allSettled`` just waits for all promises to settle, regardless of the result. The resulting array has:
```js
{status:"fulfilled", value:result} for successful responses,
{status:"rejected", reason:error} for errors.
```
```js
For example, we’d like to fetch the information about multiple users. Even if one request fails, we’re still interested in the others.

Let’s use Promise.allSettled:

let urls = [
  'https://api.github.com/users/iliakan',
  'https://api.github.com/users/remy',
  'https://no-such-url'
];

Promise.allSettled(urls.map(url => fetch(url)))
  .then(results => { // (*)
    results.forEach((result, num) => {
      if (result.status == "fulfilled") {
        alert(`${urls[num]}: ${result.value.status}`);
      }
      if (result.status == "rejected") {
        alert(`${urls[num]}: ${result.reason}`);
      }
    });
  });
```

### Promise.race
Similar to ``Promise.all``, but waits only for the first settled promise and gets its result (or error).

The syntax is:
```js
let promise = Promise.race(iterable);
```
For instance, here the result will be 1:
```js
Promise.race([
  new Promise((resolve, reject) => setTimeout(() => resolve(1), 1000)),
  new Promise((resolve, reject) => setTimeout(() => reject(new Error("Whoops!")), 2000)),
  new Promise((resolve, reject) => setTimeout(() => resolve(3), 3000))
]).then(alert); // 1
```
The first promise here was fastest, so it became the result. After the first settled promise “wins the race”, all further results/errors are ignored.

### Promise.any
Similar to ``Promise.race``, but waits only for the first fulfilled promise and gets its result. If all of the given promises are rejected, then the returned promise is rejected with AggregateError – a special error object that stores all promise errors in its errors property.

The syntax is:
```js
let promise = Promise.any(iterable);
```
For instance, here the result will be 1:
```js
Promise.any([
  new Promise((resolve, reject) => setTimeout(() => reject(new Error("Whoops!")), 1000)),
  new Promise((resolve, reject) => setTimeout(() => resolve(1), 2000)),
  new Promise((resolve, reject) => setTimeout(() => resolve(3), 3000))
]).then(alert); // 1
```
The first promise here was fastest, but it was rejected, so the second promise became the result. After the first fulfilled promise “wins the race”, all further results are ignored.

### Promise.resolve/reject
Methods ``Promise.resolve`` and ``Promise.reject`` are rarely needed in modern code, because ``async/await`` syntax (we’ll cover it a bit later) makes them somewhat obsolete.

``Promise.resolve(value)`` creates a resolved promise with the result value.

Same as:
``
let promise = new Promise(resolve => resolve(value));
``
The method is used for compatibility, when a function is expected to return a promise.
For example, the ``loadCached`` function below fetches a URL and remembers (caches) its content. For future calls with the same URL it immediately gets the previous content from cache, but uses ``Promise.resolve ``to make a promise of it, so the returned value is always a promise:
```js
let cache = new Map();

function loadCached(url) {
  if (cache.has(url)) {
    return Promise.resolve(cache.get(url)); // (*)
  }

  return fetch(url)
    .then(response => response.text())
    .then(text => {
      cache.set(url,text);
      return text;
    });
}
```
We can write ``loadCached(url).then(…)``, because the function is guaranteed to return a promise. We can always use ``.then`` after loadCached. That’s the purpose of ``Promise.resolve`` in the line (*).

- Promise.reject
``Promise.reject(error)`` creates a rejected promise with error.

Same as:
``
let promise = new Promise((resolve, reject) => reject(error));
``
In practice, this method is almost never used.